controllers/webhook/storageclass.go

```diff
@@ -0,0 +1,173 @@
+	"slices"
+
+	"github.com/red-hat-storage/ocs-operator/api/v4/v1alpha1"
+
+	"github.com/go-logr/logr"
+	"github.com/red-hat-storage/ocs-operator/v4/controllers/util"
+	jsonpatch "gomodules.xyz/jsonpatch/v2"
+	corev1 "k8s.io/api/core/v1"
+	storagev1 "k8s.io/api/storage/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+type StorageClassAdmission struct {
+	client.Client
+	Namespace string
+	Decoder   admission.Decoder
+	Log       logr.Logger
+}
+
+var supportedProvisioners = []string{
+	util.RbdDriverName,
+	util.CephFSDriverName,
+	util.NfsDriverName,
+}
+
+func (s *StorageClassAdmission) Handle(ctx context.Context, req admission.Request) admission.Response {
+	s.Log.Info("Request received for admission review")
+
+	storageClass := &storagev1.StorageClass{}
+	if err := s.Decoder.Decode(req, storageClass); err != nil {
+		s.Log.Error(err, "failed to decode admission review as storageclass")
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("only storageclasses admission reviews are supported: %v", err))
+	}
+
+	if !slices.Contains(supportedProvisioners, storageClass.Provisioner) {
+		s.Log.Error(fmt.Errorf("unsupported provisioner %s", storageClass.Provisioner), "failed validation", "storageClass", storageClass.Name)
+		return admission.Errored(http.StatusBadRequest, fmt.Errorf("supported provisioners are %s", supportedProvisioners))
+	}
+
+	storageCluster, err := util.GetStorageClusterInNamespace(ctx, s.Client, s.Namespace)
+	if err != nil {
+		s.Log.Error(err, "failed to list storagecluster", "namespace", s.Namespace)
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failed to get storagecluster in %s namespace", s.Namespace))
+	}
+
+	storageConsumerList := &v1alpha1.StorageConsumerList{}
+	if err := s.List(ctx, storageConsumerList, client.InNamespace(s.Namespace)); err != nil {
+		s.Log.Error(err, "failed to list storageconsumers", "namespace", s.Namespace)
+		return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failed to list storageconsumers in %s namespace", s.Namespace))
+	}
+	if len(storageConsumerList.Items) == 0 {
+		s.Log.Error(fmt.Errorf("no storageconsumers found in %s namespace", s.Namespace), "failed validation")
+		return admission.Denied("creation of storageclass should happen after storageconsumer exists")
+	}
+
+	var localConsumer *v1alpha1.StorageConsumer
+	for idx := range storageConsumerList.Items {
+		consumer := &storageConsumerList.Items[idx]
+		controller := metav1.GetControllerOfNoCopy(consumer)
+		if controller != nil && controller.UID == storageCluster.UID {
+			localConsumer = consumer
+			break
+		}
+	}
+	if localConsumer == nil {
+		s.Log.Error(fmt.Errorf("no storageconsumer found with storagecluster as the owner"), "failed validation")
+		return admission.Denied("creation of storageclass should happen after existence of local storageconsumer")
+	}
```

Comment:
This logic is faulty as the StorageCluster owns all the consumers, it was introduced in 4.18

Reply:
Look for ownStorageConsumersInNamespace